### PR TITLE
ip private ability

### PR DESCRIPTION
**Pull request by lcd1232**
### What does this changes

- Add optional parameter _private_ for **ipv4** and **ipv6** which gives ability to return ip address from private network

### What was wrong

You can't choose which ip do you need.

### CODE REVIEW

File: faker/providers/internet/__init__.py [hunk 1]
```diff
@@ -5,10 +5,11 @@
 
 from .. import BaseProvider
 
-from ipaddress import ip_address, ip_network, IPV4LENGTH, IPV6LENGTH
+from ipaddress import ip_address, ip_network, IPV4LENGTH, IPV6LENGTH, _IPv4Constants, _IPv6Constants
```
Comment (fcurella) [inline faker/providers/internet/__init__.py]: I don't htink we should use undocumented feature of the stdlib. They might change in the future and break `Faker`

File: faker/providers/internet/__init__.py [hunk 2]
```diff
@@ -169,20 +170,56 @@ def url(self):
         pattern = self.random_element(self.url_formats)
         return self.generator.parse(pattern)
 
-    def ipv4(self, network=False):
-        """Produce a random IPv4 address or network with a valid CIDR"""
-        address = str(ip_address(self.generator.random.randint(
-            0, (2 ** IPV4LENGTH) - 1)))
-        if network:
+    def ipv4(self, network=False, private=False, simple=True):
```
Comment (fcurella) [inline faker/providers/internet/__init__.py]: I'm -1 on `simple`. Users should not be expected to worry about how `Faker` generates the value.

Comment (ZuluPro) [reply faker/providers/internet/__init__.py]: Yes -1 on simple

**Comment by lcd1232** [comment]
@fcurella done

**Comment by sosiska** [comment]
ping @fcurella

File: faker/providers/internet/__init__.py [hunk 3]
```diff
@@ -169,20 +199,50 @@ def url(self):
...
-            0, (2 ** IPV4LENGTH) - 1)))
-        if network:
+    def ipv4(self, network=False, private=False):
+        """Produce a random IPv4 address or network with a valid CIDR."""
+        if private is True:
+            networks = dict()
+            for i in range(len(_private_ipv4_networks)):
+                networks[i] = _private_ipv4_networks[i].num_addresses
+            k = self.generator.random_element(networks)
+            _net = _private_ipv4_networks[k]
```
Comment (fcurella) [inline faker/providers/internet/__init__.py]: I'm not sure I'm following this logic.  I don't see the point of the `for` loop. Seems to me that it's picking up a random item of the `_private_ipv4_networks` list to assign to `_net` and could be done with just `_net = self.generator.random.choice(_private_ipv4_networks)`

Comment (fcurella) [reply faker/providers/internet/__init__.py]: would you put this explanation in a comment?

File: faker/providers/internet/__init__.py [hunk 4]
```diff
@@ -169,20 +199,50 @@ def url(self):
...
+            address = str(ip_address(self.generator.random.randint(
+                int(_net.network_address),
+                int(_net.network_address) + _net.num_addresses - 1
+            )))
+        else:
+            # It's unfair to trying to not find private ip but it's much faster
+            address = str(ip_address(self.generator.random.randint(
+                0, (2 ** IPV4LENGTH) - 1)))
+            while ip_address(force_text(address)).is_private is True:
```
Comment (fcurella) [inline faker/providers/internet/__init__.py]: I would default the `private` argument to `None`, and have this check only `if private is False`

**Comment by fcurella** [comment]
@lcd1232 do you mind reviewing https://github.com/joke2k/faker/pull/703 ? I think we're going to go with that implementation, since it gives us more flexibility (namely, users could choose address classes)

**Comment by lcd1232** [comment]
@fcurella yes, I'm agree with you. Should I close this PR?

**Comment by fcurella** [comment]
@lcd1232 yes, this one can be closed.

**Pull request closed without merging**